### PR TITLE
removed up & down arrows from cart

### PR DESCRIPTION
**Pull request by sojeri**
- removed up arrow from display when product has no available stock
- removed down arrow from display when order item's quantity is 1
- updated relevant order item specs

### CODE REVIEW

File: app/controllers/order_items_controller.rb [hunk 1]
```diff
@@ -4,11 +4,11 @@ class OrderItemsController < ApplicationController
   # more increases the quantity of an item in the cart
   def more # OrderItem.more <-- gimme more of this OrderItem
     if @item.product.has_available_stock?
-      @item.increment!(:quantity_ordered, 1)
+      @item.more!
     else
       # !W we should talk about whether this error message should be shown
       # and what it should truly say in production
-      flash[:error] = "You cannot increase the quantity of #{ @item.product.name } any further, because there are only #{ @item.quantity_ordered } in stock."
+      flash[:error] = @item.errors[:quantity_ordered]
```
Comment (catchingash) [inline app/controllers/order_items_controller.rb]: This would make more sense to me if it were just `@item.errors` in order to accommodate other errors that may happen.

File: app/models/order_item.rb [hunk 2]
```diff
@@ -13,12 +13,26 @@ class OrderItem < ActiveRecord::Base
...
+  # OPTIMIZE if this is only going to be used in the cart, maybe just use it inline inside the view
   def remove_prompt_text
     "Are you sure you want to remove this item (#{ product.name }) from your cart?"
   end
 
+  def more!
+    reload
+
+    # if the OrderItem isn't reloaded, this will resolve based on a cached operation
+    # by reloading, we force a new SQL query to check whether product_has_stock?
```
Comment (catchingash) [inline app/models/order_item.rb]: I'm not sure what this comment means.

File: app/models/order_item.rb [hunk 3]
```diff
@@ -13,12 +13,26 @@ class OrderItem < ActiveRecord::Base
...
-  # this should probably be in helpers/order_items_helpers
-  # if this is only going to be used in the cart, maybe just use it inline inside the view
+  # TODO this should probably be in helpers/order_items_helpers
+  # OPTIMIZE if this is only going to be used in the cart, maybe just use it inline inside the view
   def remove_prompt_text
     "Are you sure you want to remove this item (#{ product.name }) from your cart?"
   end
 
+  def more!
+    reload
```
Comment (catchingash) [inline app/models/order_item.rb]: Why is this necessary?

File: app/models/order_item.rb [hunk 4]
```diff
@@ -13,12 +13,26 @@ class OrderItem < ActiveRecord::Base
...
+    if product_has_stock?
+      increment!(:quantity_ordered, 1)
+    end
+  end
+
+  def less!
+    update_column(:quantity_ordered, quantity_ordered - 1) if quantity_ordered > 1
+  end
+
   # group: note that we need to talk about this more
```
Comment (catchingash) [inline app/models/order_item.rb]: While you're in here, do you want to add a `TODO` in this comment?

File: app/controllers/order_items_controller.rb (hunk 1 above)
Comment (sojeri) [inline app/controllers/order_items_controller.rb]: Other errors should be adding themselves to the errors hash already, no? Or maybe I don't understand what you mean.

File: app/models/order_item.rb [hunk 5]
```diff
@@ -36,12 +50,13 @@ def item_price # v. price / cost / item_cost / total_item_price / total_item_cos
 
   def product_has_stock? # !Q is this the right way to do this?
     # stock = product.has_available_stock?
-    unless product.has_available_stock?
+    if product.has_available_stock?
```
Comment (catchingash) [inline app/models/order_item.rb:59]: Does this method return true/false? I think this could be refactored so that the result of .has_available_stock? is returned, and the error message set elsewhere?

File: app/views/orders/_display_mutable_order_items.html.erb [hunk 6]
```diff
@@ -8,11 +8,14 @@
   </tr>
   <% @order.order_items.each do |item| %>
     <tr>
-      <!-- !L this should use Lindsey's display_dollars helper once it exists! -->
-      <td><!--˅˄-->
-        <%= link_to "-", less_item_path(item), method: :patch %>
+      <td>
+        <% if item.quantity_ordered > 1 %>
+          <%= link_to "-", less_item_path(item), method: :patch %>
```
Comment (catchingash) [inline app/views/orders/_display_mutable_order_items.html.erb:13]: It's fine if you prefer links, but I like to follow the rule that a link should only be a GET request, and a button should be any other kind of request.

File: spec/models/order_item_spec.rb [hunk 7]
```diff
@@ -177,6 +177,34 @@
...
+        expect(item.quantity_ordered).to eq(4)
+      end
+
+      it "unless that quantity is one or less" do
+        item = OrderItem.create(product_id: @product.id, order_id: @order.id, quantity_ordered: 1)
+        item.less!
+        expect(item.quantity_ordered).to eq(1)
+      end
+    end
+
```
Comment (catchingash) [inline spec/models/order_item_spec.rb:209]: Is there a test to confirm that the quantity cannot be increased beyond available stock, or decreased below 1?

Comment (sojeri) [inline spec/models/order_item_spec.rb:209]: yes, lines 201-205 & lines 187-191

File: app/controllers/order_items_controller.rb (hunk 1 above)
Comment (catchingash) [inline app/controllers/order_items_controller.rb]: They're not adding themselves to the flash[:error], no.

File: app/models/order_item.rb (hunk 5 above)
Comment (sojeri) [inline app/models/order_item.rb:59]: How should the error message be set elsewhere? Like-- do you have an idea where that code should live?

File: app/models/order_item.rb (hunk 3 above)
Comment (sojeri) [inline app/models/order_item.rb]: Do you understand the comments now?

File: app/models/order_item.rb [hunk 8]
```diff
@@ -13,13 +13,30 @@ class OrderItem < ActiveRecord::Base
...
+  # TODO this should probably be in helpers/order_items_helpers
+  # OPTIMIZE if this is only going to be used in the cart, maybe just use it inline inside the view
   def remove_prompt_text
     "Are you sure you want to remove this item (#{ product.name }) from your cart?"
   end
 
-  # group: note that we need to talk about this more
+  def more!
+    reload # removing this line == DANGER WILL ROBINSON
+    # if the OrderItem isn't reloaded, ln28 will resolve based on a cached operation
```
Comment (catchingash) [inline app/models/order_item.rb]: do you mean ln 30? It might be better to just reference the method by name.

File: app/models/order_item.rb [hunk 9]
```diff
@@ -13,13 +13,30 @@ class OrderItem < ActiveRecord::Base
...
+    # in other words, if line 28 has recently been evaluated for this OrderItem,
+    # then __in will not be evaluated__ and the cached value (true) will be used
```
Comment (catchingash) [inline app/models/order_item.rb]: typo? 'in will not be evaluated'?

File: app/models/order_item.rb (hunk 8 above)
Comment (sojeri) [inline app/models/order_item.rb]: whoops. it was line 28 before I expanded the comment. >_>

File: spec/models/order_item_spec.rb (hunk 7 above)
Comment (catchingash) [inline spec/models/order_item_spec.rb:209]: Perfect. Thanks! Clearly I'm being lazy today.

File: app/models/order_item.rb (hunk 5 above)
Comment (catchingash) [inline app/models/order_item.rb:59]: I like the #optimize tag you added. That seems like the best approach here. I was envisioning some sort of validation or something, rather than a before_save or whatever. But this is fine for now.

File: spec/controllers/order_items_controller_spec.rb [hunk 10]
```diff
@@ -81,6 +81,7 @@
         end
 
         expect(assigns(:item).quantity_ordered).to eq(@max_quantity - already_tied_quantity)
+        expect(assigns(:item).errors.keys).to include(:quantity_ordered)
```
Comment (catchingash) [inline spec/controllers/order_items_controller_spec.rb]: We still need a test for flash[:errors]. 

Also, making sure the errors includes quantity_ordered seems like a model test, not a controller test.
